Fix tests

neuroevolutus opened the pull request:
I think I've managed to fix a few of the tests by renaming some nodes in the `scrapscript.txt` file. I think some other tests can be considered fixed if we remove the " " character from the list of infix operators.

I tried using the the `extras` array to specify raw JavaScript regular expressions for whitespace and comments, but they don't seem to fix the comment test. Oddly enough, what does seem to work is putting references to unnamed rules for whitespace and comments within the `extras` array that return regular expressions with the exact same content as the regular expressions I tried putting directly in the array previously.

With regards to the "where" clause, are semicolons being used to separate all declarations as shown below? That is to say, we're not using the semicolon to just delimit the start of a set of declarations?

```
f 1 && g 2 ; f = | 1 -> #true () ; g = | 2 -> #false ()
```

surprisetalk commented:
Looks great! Thank youuu

> `f 1 && g 2 ; f = | 1 -> #true () ; g = | 2 -> #false ()`

Yep, that example is correct.

Another example: `a + b + c ; a = 1 ; b = 2 ; c = 3`

Another example: `x : t = 1 ; y : t = 2 ; t : int`

neuroevolutus commented:
Sweet! Just to bounce off that and clarify, is it indeed possible to have no top-level expression and just a sequence of declarations as in your second example (or can the first declaration in a sequence of declarations counts as an expression)?

> `x : t = 1 ; y : t = 2 ; t : int`

Continuing off that, if we have to have at least one top-level expression, do we only ever have _at most_ one top-level expression?

> `a + b + c ; a = 1 ; b = 2 ; c = 3`

Third, in "fun TEST 1", we seem to end the function with a `|` despite not following it with another match arm. Is this intentional as an optional function delimiter?

surprisetalk commented:
> Sweet! Just to bounce off that and clarify, is it indeed possible to have no top-level expression and just a sequence of declarations as in your second example (or can the first declaration in a sequence of declarations counts as an expression)?

`a ; a = 1` is an expression. You can use it like `b ; b = (a ; a = 1)`.

You can mostly treat`;` like a right-associative operator:`1 ; a = 2 ; b = 3` is equivalent to `(((1) ; a = 2) ; b = 3)`.

But `a = 1` and `a ; 1` are not valid expressions.

> Continuing off that, if we have to have at least one top-level expression, do we only ever have at most one top-level expression?

Every scrapscript program is exactly one top-level expression.

There are no imports, nor headers, nor module statements, nor type declarations -- everything is part of the top-level expression.

> Third, in "fun TEST 1", we seem to end the function with a | despite not following it with another match arm. Is this intentional as an optional function delimiter?

Whoops -- that sounds like a typo. We should never have a trailing pipe character!

neuroevolutus commented:
Thanks, Taylor. I think that clears things up for me.

neuroevolutus commented:
I've got all the existing tests passing! I made heavy changes to `test/corpus/scrapscript.txt`, but I'm hoping the new ASTs are still sensible. I have some additional doubts about other parts of the grammar we haven't exercised yet, but I'm wondering if we can move those into a another issue / PR.

A few things I wanted to note for this PR:

1. I added support for record field puns (e.g. "fun TEST 9").
2. I removed the extra `$._expr` production from the `...` rule (e.g. "fun TEST 7").
3. I removed `|` as a binary operator since the pipe was being handled in a match function production. I made the assumption that bitwise OR was not needed.

I just wanted to make sure these details in particular were not off base on my part.

I'm looking forward to hearing your thoughts!
